lua/wire/wirenet.lua

```diff
@@ -207,4 +207,226 @@ function Net.Trivial.Receive(name, callback)
+	local written = cb()
+	if written>0 then
+		ss.index = sizepos
+		ss:writeInt32(written)
+		ss.index = #ss+1
+		ss.subindex = 1
+	else
+		ss.index = sizepos
+		for i=sizepos, #ss do
+			ss[i] = nil
+		end
+	end
+	return written
+end
+local function calcWritten(ss, startpos)
+	local size = 0
+	for i=startpos, ss.index-1 do
+		size = size + #ss[i]
+	end
+	return size
+end
+
+local WireMemSyncer = {
+	__index = {
+		register = function(self, ent)
+			self.entities[ent] = true
+		end,
+		sync = function(self)
+			if not self.syncing then
+				self.syncing = true
+				timer.Simple(0.5, function() self.syncing = false self:dosync() end)
+			end
+		end,
+		dosync = function(self)
+			if self.sending then self:sync() return end
+			local ss = StringStream()
+			local sizeleft = net.Stream.SendSize*net.Stream.MaxServerChunks*0.5
+
+			if writeUnboundedArray(ss, function()
+				local numEntries = 0
+				for ent in pairs(self.entities) do
+					if ent:IsValid() then
+						if ent.MemoryDirty then
+
+							ss:writeInt16(ent:EntIndex())
+							local written = writeUnboundedArray(ss, function()
+								local startpos = ss.index
+								ent:SerializeMemory(ss, sizeleft)
+								return calcWritten(ss, startpos)
+							end)
+
+							if written>0 then
+								sizeleft = sizeleft - written - 6
+								numEntries = numEntries + 1
+							else
+								-- Delete the entIndex
+								ss.index = ss.index - 1
+								ss[ss.index] = nil
+							end
+
+							if sizeleft <= 0 then self:sync() break end
+						end
+					else
+						self.entities[ent] = nil
+					end
+				end
+				return numEntries
+			end) > 0 then
+				self:send(ss:getString())
+			end
+
+		end,
+		fullsnapshot = function(self, ply)
+			local snapshots = {}
+			local curent = nil
+
+			while true do
+				local shouldBreak = true
+				local ss = StringStream()
+				local sizeleft = net.Stream.SendSize*net.Stream.MaxServerChunks*0.5
+
+				if writeUnboundedArray(ss, function()
+					local numEntries = 0
+					for ent, mem in next, self.entities, curent do
+						if ent:IsValid() then
+
+							ss:writeInt16(ent:EntIndex())
+							local written = writeUnboundedArray(ss, function()
+								local startpos = ss.index
+								ent:SerializeFullMemory(ss, sizeleft)
+								return calcWritten(ss, startpos)
+							end)
+
+							if written>0 then
+								sizeleft = sizeleft - written - 6
+								numEntries = numEntries + 1
+								if sizeleft <= 0 then
+									shouldBreak = false
+									curent = ent
+									break
+								end
+							else
+								-- Delete the entIndex
+								ss.index = ss.index - 1
+								ss[ss.index] = nil
+							end
+
+						else
+							self.entities[ent] = nil
+						end
+					end
+					return numEntries
+				end) > 0 then
+					snapshots[#snapshots + 1] = ss:getString()
+				end
+				if shouldBreak then break end
+			end
+
+			for _, v in ipairs(snapshots) do
+				self:send(v, ply, #snapshots)
+			end
+		end,
+		send = function(self, data, ply, snapshots)
```

Comment:
Should be two different functions instead of using the ply var to seperate this one into two branches.